RFC0003-sparse-roadmap/README.md

```diff
@@ -0,0 +1,1352 @@
+with multiple dispatch that in turn has derivatives defined on it.
+
+##### Method calls on COO sparse tensors
+
+The internal implementation of COO tensors is done within the `SparseTensorImpl` class defined
+in [SparseTensorImpl.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseTensorImpl.cpp).
+This is a subclass of `TensorImpl` that specialize it for COO sparse
+by including some extra members and functions found in [SparseTensorImpl.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseTensorImpl.h).
+
+##### Sparse method dispatch
+
+The `dispatch` parameter that is specified for functions written in `native_functions.yaml` can also accept
+a dispatch parameter of type `SparseCPU` or `SparseCUDA` that allows listing functions that are specifically
+targeted at sparse data layout. Notable examples of such functions can be found in `torch.add()` that defines
+different dispatch calls for different device and layout types. The impact on CSR sparse can be that we can
+make additions to code generators in
+[preprocess_declarations.py](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/preprocess_declarations.py)
+and add another `SparseCSRCPU` dispatch
+backend and write specialized function for any method just by adding a new entry into `native_functions.yaml`.
+
+### CRS/CCS sparse array storage format
+
+The [CRS/CCS](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_(CSR,_CRS_or_Yale_format)) sparse array storage format has a long history (since
+mid-1960s) for representing sparse matrices because this format is memory
+efficient and convinient for multiplying sparse matrices.  However,
+the CRS/CCS format was designed for storing sparse two-dimensional
+arrays. A number of generalizations of CRS/CCS format exists that
+makes the choice of generalization for PyTorch not as straightforward
+as one could wish.
+
+Here we propose to use our modification of [Generalized Compressed
+Row/Column Storage format
+(GCRS/GCCS)](https://ieeexplore.ieee.org/document/7237032). The basic
+idea behind GCRS/GCCS is as follows:
+- a multidimensional array is represented as two-dimensional array
+  using [dimensionality reduction](#dimensionality-reduction),
+- the associated two-dimensional array is stored using the standard
+  Compressed Row/Column Storage format (CRS/CCS).
+
+The original version of GCRS/GCCS format uses a specific choice of
+dimensionality reduction where the dimensions are split into two parts
+using the evenness of dimensions numbers. This choice does not seem to
+have obvious advantages over other possible choises from the view
+point of data locality and performance. So, our modification will be
+parameterized verion of GCRS/GCCS where the choice of splitting the
+set of dimensions to two sets can be user's choice. Also, this
+approach allows one implementation for both CRS and CCS format cases.
```

Comment:
@v0dro Parameterizing GCRS/GCCS doesn't seem like an unambiguous win. In particular, whenever you parametrize in this way, you have to record extra metadata regarding what parametrization was chosen. If there is a fixed convention (like the original paper's even/odd), you no longer need to store this data. Similarly, it's not obvious to me that one implementation supporting both CRS and CCS is actually a good thing! It certainly makes life a lot harder for prospective kernel writers.

I am a little uncertain whether or not parametrized form makes the indexing formulas more complicated or not because arbitrary rank indexing is difficult no matter what you do (since it's a loop, not a fixed formula).